Lib/gzip.py

```diff
@@ -580,43 +580,21 @@ def _rewind(self):
         self._new_member = True
 
 
-def _create_simple_gzip_header(compresslevel: int,
-                               mtime = None) -> bytes:
-    """
-    Write a simple gzip header with no extra fields.
-    :param compresslevel: Compresslevel used to determine the xfl bytes.
-    :param mtime: The mtime (must support conversion to a 32-bit integer).
-    :return: A bytes object representing the gzip header.
-    """
-    if mtime is None:
-        mtime = time.time()
-    if compresslevel == _COMPRESS_LEVEL_BEST:
-        xfl = 2
-    elif compresslevel == _COMPRESS_LEVEL_FAST:
-        xfl = 4
-    else:
-        xfl = 0
-    # Pack ID1 and ID2 magic bytes, method (8=deflate), header flags (no extra
-    # fields added to header), mtime, xfl and os (255 for unknown OS).
-    return struct.pack("<BBBBLBB", 0x1f, 0x8b, 8, 0, int(mtime), xfl, 255)
-
-
 def compress(data, compresslevel=_COMPRESS_LEVEL_BEST, *, mtime=None):
     """Compress data in one shot and return the compressed string.
 
     compresslevel sets the compression level in range of 0-9.
     mtime can be used to set the modification time. The modification time is
     set to the current time by default.
     """
-    if mtime == 0:
-        # Use zlib as it creates the header with 0 mtime by default.
-        # This is faster and with less overhead.
-        return zlib.compress(data, level=compresslevel, wbits=31)
-    header = _create_simple_gzip_header(compresslevel, mtime)
-    trailer = struct.pack("<LL", zlib.crc32(data), (len(data) & 0xffffffff))
-    # Wbits=-15 creates a raw deflate block.
-    return (header + zlib.compress(data, level=compresslevel, wbits=-15) +
-            trailer)
+    # Wbits=31 automatically includes a gzip header and trailer.
+    gzip_data = zlib.compress(data, level=compresslevel, wbits=31)
+    if mtime is None:
+        mtime = time.time()
+    # Reuse gzip header created by zlib, replace mtime and OS byte for
+    # consistency.
+    header = struct.pack("<4sLBB", gzip_data, int(mtime), gzip_data[8], 255)
+    return header + gzip_data[10:]
```

Comment:
This causes a large data copy and requires double the memory.  We really should do this header rewriting on the original bytes buffer in C within zlib.compress, otherwise it is a performance regression.

Reply:
Regression is the wrong word, the old code this replaces did an equivalent copy... we could just do better.  Given the old code's behavior, don't let this comment block the PR. :)

Reply:
The old code did two copies.

Also, I suppose that the cost of copying is insignificant in comparison with the cost of compressing. It may be not worth to spent our time on it.

Reply:
The new code also does two copies, so it is the same.

Reply:
The cost of copying is insignificant to the cost of compression. Using zlib.compress, which writes one buffer, is barely faster. At most about 1-2% or so. This is not worth optimizing over IMO.

Reply:
> We really should do this header rewriting on the original bytes buffer in C within zlib.compress, otherwise it is a performance regression.

I did one suggestion earlier about improving gzip earlier, but we decided not to go through because of PEP-399 (https://github.com/python/cpython/issues/110283#issuecomment-1910354820). That states that an equivalent python option must always remain present. So moving it to C requires then to have two equivalent code paths to maintain. That is not really worth the 1% performance upgrade.

Reply:
The cost is unnecessary double memory use.  /But did you see where I said don't worry about this?

The optimization **is** worthwhile.  But I don't want it done in this PR, if it happens at all, it'll probably be a followup from me. :)  --  PEP-399 does not apply because zlib is not implemented in Python and never will be. The header bytes modification would be done in the `zlib.compress` extension code.

Reply:
I was not aware that PEP-399 did not apply.
As a sidenote, I think more optimally estimating initial memory use using the deflatebound functions is also important as PyMem_Realloc also does a new allocation plus a copy in some occassions, which causes unecessary memory usage.